File: pkg/promtail/client/batch.go
package client

import (
	"time"

	"github.com/gogo/protobuf/proto"
	"github.com/golang/snappy"
	"github.com/grafana/loki/pkg/logproto"
	"github.com/prometheus/common/model"
)

// batch holds pending log streams waiting to be sent to Loki, and it's used
// to reduce the number of push requests to Loki aggregating multiple log streams
// and entries in a single batch request. In case of multi-tenant Promtail, log
// streams for each tenant are stored in a dedicated batch.
type batch struct {
	streams   map[model.Fingerprint]*logproto.Stream
	bytes     int
	createdAt time.Time
}

func newBatch(entries ...entry) *batch {
	b := &batch{
		streams:   map[model.Fingerprint]*logproto.Stream{},
		bytes:     0,
		createdAt: time.Now(),
	}

	// Add entries to the batch
	for _, entry := range entries {
		b.add(entry)
	}

	return b
}

// add an entry to the batch
func (b *batch) add(entry entry) {
	b.bytes += len(entry.Line)

	// Append the entry to an already existing stream (if any)
	fp := entry.labels.FastFingerprint()
	if stream, ok := b.streams[fp]; ok {
		stream.Entries = append(stream.Entries, entry.Entry)
		return
	}

	// Add the entry as a new stream
	b.streams[fp] = &logproto.Stream{
		Labels:  entry.labels.String(),
		Entries: []logproto.Entry{entry.Entry},
	}
}

// sizeBytes returns the current batch size in bytes
func (b *batch) sizeBytes() int {
	return b.bytes
}

// sizeBytesAfter returns the size of the batch after the input entry
// will be added to the batch itself
func (b *batch) sizeBytesAfter(entry entry) int {
	return b.bytes + len(entry.Line)
}

// age of the batch since its creation
func (b *batch) age() time.Duration {
	return time.Since(b.createdAt)
}

// encode the batch as snappy-compressed push request, and returns
// the encoded bytes and the number of encoded entries
func (b *batch) encode() ([]byte, int, error) {
	req := logproto.PushRequest{
		Streams: make([]*logproto.Stream, 0, len(b.streams)),
	}

	entriesCount := 0
	for _, stream := range b.streams {
		req.Streams = append(req.Streams, stream)
		entriesCount += len(stream.Entries)
	}

	buf, err := proto.Marshal(&req)
	if err != nil {
		return nil, 0, err
	}
	buf = snappy.Encode(nil, buf)
	return buf, entriesCount, nil
}

File: cmd/logstash/lib/logstash/outputs/loki/batch.rb
require 'time'

module LogStash
    module Outputs
        class Loki 
            class Batch
                attr_reader :streams
                def initialize(e)
                    @bytes = 0
                    @createdAt = Time.now
                    @streams = {}
                    add(e)
                end 

                def size_bytes()
                    return @bytes
                end

                def add(e)
                    @bytes = @bytes + e.entry['line'].length

                    # Append the entry to an already existing stream (if any)
                    labels = e.labels.to_s
                    if @streams.has_key?(labels)
                        stream = @streams[labels]
                        stream['entries'] = stream['entries'] + e.entry
                        return
                    else  
                        # Add the entry as a new stream
                        @streams[labels] = {
                            "labels" => e.labels,
                            "entries" => [e.entry],
                        }
                    end
                end 

                def size_bytes_after(line)
                    return @bytes + line.length
                end 

                def age()
                    return Time.now - @createdAt
                end
            end 
        end 
    end 
end 